pkg/operator/encryption/state.go

```diff
@@ -43,175 +45,241 @@ func getCurrentEncryptionConfig(secrets corev1client.SecretInterface, revision s
-	})
-
-	// this is our output from the for loop below
-	encryptionState := groupResourcesState{}
-
-	// TODO if we cannot find the secrets associated with the keys, they may have been deleted
-	// we could try to use the key data in the revision to recover or just go degraded permanently
-	for gr, keys := range getGRsActualKeys(encryptionConfig) {
-		writeSecret := findSecretForKey(keys.writeKey, encryptionSecrets, targetNamespace) // TODO handle nil as error when hasWriteKey == true?
-		readSecrets := make([]*corev1.Secret, 0, len(keys.readKeys)+1)
-		if keys.hasWriteKey() {
-			readSecrets = append(readSecrets, writeSecret)
-		}
-		for _, readKey := range keys.readKeys {
-			readSecret := findSecretForKey(readKey, encryptionSecrets, targetNamespace) // TODO handle nil as error?
-			readSecrets = append(readSecrets, readSecret)
-		}
-
-		encryptionState[gr] = keysState{
-			targetNamespace: targetNamespace,
-			writeSecret:     writeSecret,
-			readSecrets:     readSecrets,
+		} else {
+			encryptionSecrets = append(encryptionSecrets, s.DeepCopy()) // don't use &s which is constant in the loop
 		}
 	}
+	encryptionSecrets = sortRecentFirst(encryptionSecrets)
 
-	// see if any resource is missing and properly reflect that state so that
-	// we will add the read keys and take it through the transition correctly.
-	for gr := range encryptedGRs {
-		if _, ok := encryptionState[gr]; !ok {
-			encryptionState[gr] = keysState{
+	// add new resources without keys. These resources will trigger STEP 2.
+	for gr := range toBeEncryptedGRs {
+		if _, ok := desiredEncryptionState[gr]; !ok {
+			desiredEncryptionState[gr] = keysState{
 				targetNamespace: targetNamespace,
 			}
 		}
 	}
 
-	// TODO allow removing resources. This would require transitioning back to identity.  For now, because we work
-	// against a single key, we can simply keep moving that additional resource even in the face of downgrades.
+	//
+	// STEP 1: without secrets, wait for the key controller to create one
+	//
+	// the code after this point assumes at least one secret
+	if len(encryptionSecrets) == 0 {
+		klog.V(4).Infof("no encryption secrets found")
+		return desiredEncryptionState
+	}
 
+	//
+	// STEP 2: verify to have all necessary read-keys. If not, deploy and wait for stability.
+	//
+	// Note: only keysWithPotentiallyPersistedData are considered. There might be more which are not pruned yet by the pruning controller.
+	//
+	// TODO: allow removing resources (e.g. on downgrades) and transition back to identity.
+	// TODO: if the operand namespace gets deleted, this code causes us to go through a cycle of identity as the write key
 	allReadSecretsAsExpected := true
+	currentlyEncryptedGRs := make([]schema.GroupResource, 0, len(desiredEncryptionState))
+	for gr := range getGRsActualKeys(oldEncryptionConfig) {
+		currentlyEncryptedGRs = append(currentlyEncryptedGRs, gr)
+	}
+	expectedReadSecrets := keysWithPotentiallyPersistedData(currentlyEncryptedGRs, encryptionSecrets)
```

Comment:
okay, so this is an improvement, we take all the keys up that haven't been used for migration and additionally the one that has been used.